mutation_reader.cc

```diff
@@ -303,6 +320,72 @@ struct mutation_reader_merger::fragment_heap_compare {
     }
 };
 
+void mutation_reader_merger::maybe_add_readers_at_partition_boundary() {
+    // We are either crossing partition boundary or ran out of
+    // readers. If there are halted readers then we are just
+    // waiting for a fast-forward so there is nothing to do.
+    if (_fragment_heap.empty() && _halted_readers.empty()) {
+        if (_reader_heap.empty()) {
+            maybe_add_readers(std::nullopt);
+        } else {
+            maybe_add_readers(_reader_heap.front().fragment.as_partition_start().key());
+        }
+    }
+}
+
+void mutation_reader_merger::retire_reader(reader_and_last_fragment_kind rk) {
+    if (_fwd_sm == streamed_mutation::forwarding::yes && rk.last_kind != mutation_fragment::kind::partition_end) {
+        // When in streamed_mutation::forwarding mode we need
+        // to keep track of readers that returned
+        // end-of-stream to know what readers to ff. We can't
+        // just ff all readers as we might drop fragments from
+        // partitions we haven't even read yet.
+        // Readers whoose last emitted fragment was a partition
+        // end are out of data for good for the current range.
+        _halted_readers.push_back(std::move(rk));
+    } else if (_fwd_mr == mutation_reader::forwarding::no) {
+        _to_remove.splice(_to_remove.end(), _all_readers, rk.reader);
+        if (_to_remove.size() >= 4) {
+            _to_remove.clear();
+        }
+    }
+}
+
+bool mutation_reader_merger::in_gallop_mode() const {
+    return _gallop_mode_hits >= gallop_mode_entering_threshold;
+}
+
+future<stop_iteration> mutation_reader_merger::advance_galloping_reader(db::timeout_clock::time_point timeout) {
+    // When _galloping_reader is present, we assume that it will keep producing fragments
+    // that win over other readers. If that assumption turns out to be no longer true,
+    // this function exits gallop mode and we should proceed with standard merging logic.
+    return (*_galloping_reader.reader)(timeout).then([this] (mutation_fragment_opt mfo) {
+        if (mfo) {
+            if (mfo->is_partition_start()) {
+                // Don't assume we will keep winning in the next partition.
+                _reader_heap.emplace_back(_galloping_reader.reader, std::move(*mfo));
+                boost::push_heap(_reader_heap, reader_heap_compare(*_schema));
+                maybe_add_readers_at_partition_boundary();
+            } else {
+                if (_fragment_heap.empty() || position_in_partition::less_compare(*_schema)(mfo->position(), _fragment_heap.front().fragment.position())) {
```

Comment:
It appears to me that this condition is the meat of this optimization, other than that there is not much difference compared to the standard logic.
I think we could merge all three modes (multi-reader, single-reader and gallop) by just specializing `prepare_next()` to single and multiple readers by just looking at how many readers are in `_next`. When there is just one reader we try to see if we can bypass the merging logic altogether by using this very same condition. When there are multiple readers we fall back to the standard logic. We can get away without any separate `_single_reader` and `_gallopping_reader` and friends.

Reply:
`prepare_next()` would return `future<needs_merge>`. When `needs_merge` == false, `operator()()` bypasses the merging logic.

Reply:
Reading the old discussion, I can see that Paweł kept `_single_reader` logic outside the `prepare_next` function in order to skip wrapping in a future in a general case (e.g. when single reader's internal buffer still has some fragments). I will try merging `advance_galloping_reader` with `prepare_next` and will see how the performance changes.

Reply:
I experimented once optimizing for the case where the readers' buffers are not empty, avoding futures. I saw no improvements whatsoever in the our macrobenchmarks.

Reply:
(which doesn't mean you shouldn't check to see if the performance changes)

Reply:
> Reading the old discussion, I can see that Paweł kept `_single_reader` logic outside the `prepare_next` function in order to skip wrapping in a future in a general case (e.g. when single reader's internal buffer still has some fragments).

Wrapping something in a future is of course more expensive then not having to do it at all, but that is not the case here. We have to handle the case where the buffer is empty anyway, so the compiler has to generate the future handling code anyway, we are not saving anything. Furthermore calling `then()` on a ready future is actually pretty cheap, the continuation lambda is called right away. What is expensive is a future that defers as that causes task objects to be allocated for each continuation. So, as long as we avoid unnecessary defers we should be fine performance wise.